routing.go

```diff
@@ -121,111 +121,200 @@ func (dht *IpfsDHT) GetValue(ctx context.Context, key string, opts ...ropts.Opti
-			go func(v RecvdVal) {
-				if v.From == dht.self {
-					err := dht.putLocal(key, fixupRec)
-					if err != nil {
-						log.Error("Error correcting local dht entry:", err)
-					}
+	valCh, err := dht.getValues(ctx, key, responsesNeeded)
+	if err != nil {
+		return nil, err
+	}
+
+	out := make(chan []byte)
+	go func() {
+		defer close(out)
+
+		maxVals := responsesNeeded
+		if maxVals < 0 {
+			maxVals = defaultQuorum * 4 // we want some upper bound on how
+			// much correctional entries we will send
+		}
+		vals := make([]RecvdVal, 0, maxVals)
+		best := -1
+
+		defer func() {
+			if len(vals) <= 1 || best < 0 {
+				return
+			}
+			fixupRec := record.MakePutRecord(key, vals[best].Val)
+			for _, v := range vals {
+				// if someone sent us a different 'less-valid' record, lets correct them
+				if !bytes.Equal(v.Val, vals[best].Val) {
+					go func(v RecvdVal) {
+						if v.From == dht.self {
+							err := dht.putLocal(key, fixupRec)
+							if err != nil {
+								log.Error("Error correcting local dht entry:", err)
+							}
+							return
+						}
+						ctx, cancel := context.WithTimeout(dht.Context(), time.Second*30)
+						defer cancel()
+						err := dht.putValueToPeer(ctx, v.From, key, fixupRec)
+						if err != nil {
+							log.Debug("Error correcting DHT entry: ", err)
+						}
+					}(v)
+				}
+			}
+		}()
+
+		for {
+			select {
+			case v, ok := <-valCh:
+				if !ok {
 					return
 				}
-				ctx, cancel := context.WithTimeout(dht.Context(), time.Second*30)
-				defer cancel()
-				err := dht.putValueToPeer(ctx, v.From, key, fixupRec)
-				if err != nil {
-					log.Debug("Error correcting DHT entry: ", err)
+
+				i := len(vals)
+				if len(vals) < maxVals {
+					vals = append(vals, v)
+				} else {
+					i = (best + 1) % maxVals
+					vals[i] = v
 				}
-			}(v)
+
+				if v.Val == nil {
+					continue
+				}
+				// Select best value
+				if best > -1 {
+					sel, err := dht.Validator.Select(key, [][]byte{vals[best].Val, v.Val})
+					if err != nil {
+						log.Warning("Failed to select dht key: ", err)
+						continue
+					}
+					if sel == 1 && !bytes.Equal(v.Val, vals[best].Val) {
+						best = i
+						out <- v.Val
+					}
+				} else {
+					// Output first valid value
+					if err := dht.Validator.Validate(key, v.Val); err == nil {
+						best = len(vals) - 1
+						out <- v.Val
+					}
+				}
+			case <-ctx.Done():
+				return
+			}
 		}
-	}
+	}()
 
-	return best, nil
+	return out, nil
 }
 
 // GetValues gets nvals values corresponding to the given key.
 func (dht *IpfsDHT) GetValues(ctx context.Context, key string, nvals int) (_ []RecvdVal, err error) {
+	valCh, err := dht.getValues(ctx, key, nvals)
+	if err != nil {
+		return nil, err
+	}
+
+	out := make([]RecvdVal, 0, nvals)
+	for val := range valCh {
+		out = append(out, val)
+	}
+
+	return out, ctx.Err()
+}
+
+func (dht *IpfsDHT) getValues(ctx context.Context, key string, nvals int) (<-chan RecvdVal, error) {
 	eip := log.EventBegin(ctx, "GetValues")
```

Comment:
This should probably go into `GetValues`. That may also help simplify/get rid of the `done` dance.